src/aapns/connection.py

```diff
@@ -0,0 +1,440 @@
+import json
+import ssl
+from asyncio import (
+    CancelledError,
+    Event,
+    Future,
+    TimeoutError,
+    create_task,
+    open_connection,
+    wait_for,
+)
+from contextlib import suppress
+from dataclasses import dataclass, field
+from logging import getLogger
+from math import inf
+from time import time
+from typing import Dict, List, Optional
+from urllib.parse import urlparse
+
+import h2.config
+import h2.connection
+import h2.settings
+
+# Apple limits APN payload (data) to 4KB or 5KB, depending.
+# Request header is not subject to flow control in HTTP/2
+# Data is subject to framing and padding, but those are minor.
+MAX_NOTIFICATION_PAYLOAD_SIZE = 5120
+REQUIRED_FREE_SPACE = 6000
+# OK response is empty
+# Error response is short json, ~30 bytes in size
+MAX_RESPONSE_SIZE = 2 ** 16
+# Inbound connection flow control window
+# It's quite arbitrary, guided by:
+# * concurrent requests limit, server limit being 1000 today
+# * expected response size, see above
+CONNECTION_WINDOW_SIZE = 2 ** 24
+logger = getLogger(__package__)
+
+
+@dataclass
+class Channel:
+    sid: int
+    fut: Optional[Future]
+    ev: List[h2.events.Event] = field(default_factory=list)
+    header: Optional[dict] = None
+    body: bytes = b""
+
+
+class Connection:
+    # FIXME document connection termination states:
+    # * not closing: active connection
+    # * closing and not closed: graceful shutdown
+    # * closed: totally dead
+    # * fixme: __aexit__() finished
+    closed = closing = False
+    outcome: str = None
+    r = w = bgr = bgw = None
+    channels: Dict[int, Channel]
+    max_concurrent_streams = 100  # recommended in RFC7540#section-6.5.2
+    last_new_sid = last_sent_sid = -1  # client streams are odd
+
+    def __repr__(self):
+        bits = [self.state, f"{self.host}:{self.port}"]
+        if self.state != "closed":
+            bits.extend([f"buffered:{self.buffered}", f"inflight:{self.inflight}"])
+        else:
+            bits.append(f"outcome:{self.outcome}")
+        return "<Connection %s>" % " ".join(bits)
+
+    def __init__(self, origin: str, ssl=None):
+        self.channels = dict()
+        url = urlparse(origin)
+        assert url.scheme == "https"
+        assert url.hostname
+        assert not url.username
+        assert not url.password
+        assert not url.path
+        assert not url.params
+        assert not url.query
+        assert not url.fragment
+        self.host = url.hostname
+        self.port = url.port or 443
+        self.ssl = ssl if ssl else create_ssl_context()
+
+    async def __aenter__(self):
```

Comment:
This would be harder to abuse/easier to read if it was `await Connection.create(origin: str, ssl:Optional[SSLContext]) `and `Connection.__init__` getting the values that `__aenter__` creates, so it's signature would be something like `Connection.__init__(please_write: Event, reader: Reader, writer: Writer, connection: H2Connection)` (heck, make it a dataclass at that point).